distributed/worker.py

```diff
@@ -2061,121 +2387,69 @@ def story(self, *keys):
         ]
 
     def ensure_communicating(self):
-        changed = True
-        try:
-            while (
-                changed
-                and self.data_needed
-                and len(self.in_flight_workers) < self.total_out_connections
-            ):
-                changed = False
-                logger.debug(
-                    "Ensure communicating.  Pending: %d.  Connections: %d/%d",
-                    len(self.data_needed),
-                    len(self.in_flight_workers),
-                    self.total_out_connections,
-                )
-
-                key = self.data_needed[0]
-
-                if key not in self.tasks:
-                    self.data_needed.popleft()
-                    changed = True
-                    continue
-
-                ts = self.tasks[key]
-                if ts.state != "waiting":
-                    self.log.append((key, "communication pass"))
-                    self.data_needed.popleft()
-                    changed = True
-                    continue
+        stimulus_id = f"ensure-communicating-{uuid.uuid4()}"
+        skipped_worker_in_flight = list()
+        while self.data_needed and (
+            len(self.in_flight_workers) < self.total_out_connections
+            or self.comm_nbytes < self.total_comm_nbytes
+        ):
```

Comment:
The logic is now (subjectively) a bit simpler. `self.data_needed` no longer holds the dependents for which we are fetching dependencies but now it includes the actual dependencies / replicas to fetch